staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go

```diff
@@ -601,6 +704,53 @@ func (r *crdHandler) GetCustomResourceListerCollectionDeleter(crd *apiextensions
 	return info.storages[info.storageVersion].CustomResource, nil
 }
 
+// getOrCreateServingInfoForWrite is getOrCreateServingInfoFor, but blocks CR
+// write requests until the storage version update is processed, aborted, or
+// timed-out.
+// If the storage version update is processed, unblock and allow the write request.
+// If the storage version update is aborted, unblock and reject the write request.
+// If the storage version update timed-out, unblock and allow the write request.
+//
+// If aborted is true, the returned error (403) should be surfaced to the client,
+// indicating that the server forbids the CR request due to the CRD deletion.
+// Otherwise, any non-nil error returned by getOrCreateServingInfoFor may be
+// wrapped by the server as an internal error (500) before surfacing to the client.
+func (r *crdHandler) getOrCreateServingInfoForWrite(uid types.UID, name string) (ret *crdInfo, aborted bool, err error) {
+	ret, err = r.getOrCreateServingInfoFor(uid, name)
+
+	// Surface non-nil error early.
+	if err != nil ||
+		// Return early if the StorageVersionAPI feature gate is disabled.
+		!utilfeature.DefaultFeatureGate.Enabled(features.StorageVersionAPI) ||
+		!utilfeature.DefaultFeatureGate.Enabled(features.APIServerIdentity) ||
+		ret.storageVersionUpdate == nil ||
+		ret.storageVersionUpdate.processedCh == nil {
+		return ret, false, err
+	}
+
+	// NOTE: currently the graceful CRD deletion waits 1s for in-flight requests
+	// to register themselves to the wait group. Ideally the storage version update should
+	// not cause the requests to miss the 1s window; otherwise the requests may
+	// fail ungracefully (e.g. it may happen if the CRD was deleted immediately after the
+	// first CR request establishes the underlying storage).
+	select {
+	case <-ret.storageVersionUpdate.processedCh:
+	case <-ret.storageVersionUpdate.abortedCh:
+		// Storage version update aborted. The caller knows which error
+		// to return to the client (403).
```

Comment:
* how does a server/user get out of this error state?
* 403 does not seem like the correct error code... would "unavailable" or something similar be better?

Reply:
> how does a server/user get out of this error state

the error state is transient and only for in-flight requests. A CR client can retry: 
- if the CRD is gone forever, the client will get a 404. 
- if the CRD gets re-created, things continue to work

> would "unavailable" or something similar be better

had 503 originally (ref https://github.com/kubernetes/kubernetes/pull/96403#discussion_r559328435). I'm okay with 503. If we assume a CR client and a CRD client are the same user, then 403 (let the user fix the request/situation) also makes sense